### ISSUE TITLE
Assign a namespace to the tenant from the tenant prefix used in the namespace

### ISSUE THREAD

**Issue by bsctl** [clastix/capsule#123]
# Describe the feature
When `--force-tenant-prefix` is set, multiple tenants are assigned to the same owner, and the user creates a new namespace, capsule should infer the tenant to assign from the tenant prefix used in the namespace. Before the recent rework introduced by #118, this was the behaviour. 

Currently, when multiple tenants are assigned to the same owner, the user must specify the tenant with the label `capsule.clastix.io/tenant=<desired_tenant>` otherwise capsule denies with the following message `Unable to assign namespace to tenant. Please use capsule.clastix.io/tenant label when creating a namespace`.

From the UX point of view, the new behaviour looks worse. A user having more than one tenant is expecting the system to assign the namespace to the right tenant because the user is following the rule that the system is requiring. He should not be forced to add a label in the namespace to indicate the tenant he wants to be assigned. The tenant is already indicated into the name of the namespace. Forcing the user to specify it again, it looks a poor UX.

It would be nice to revert this useful feature for a better user experience.

# What would the new user story look like?
1. the capsule controller is started with `--force-tenant-prefix` by the cluster admin
2. the tenant owner gets assigned more than one tenant, eg. `oil` and `gas`
3. the tenant owner is aware of the tenant prefix rule, so she just creates a new namespace `oil-production`
4. capsule infers the desired tenant from the namespace prefix and assign `oil-production` to the `oil` tenant

# Expected behavior
When `--force-tenant-prefix` is set, Capsule infers the desired tenant from the namespace prefix and assign a new namespace to the right tenant

**Comment by MaxFedotov** [comment]
Hi @bsctl, let me try to explain why we had to introduce such a design.
According to the latest issues, users asked about the support of tenants, whose names had dashes. 
That introduces the following problem - let's assume that the user creates a namespace, with a name of `teamalpha-production-mycoolproject-web` and we had `--force-tenant-prefix` enabled.
What we do next - we take this user and his group and build a list of tenants, to whom this user had an access.
And in this list, we had following tenants:
- `teamalpha`
- `teamalpha-production`
- `teamalpha-production-mycoolproject`

So here is where the problem arises - to which tenant we should assign this namespace? 
One solution is to iterate through a list of tenants, compare namespace name with each of the tenants names and calculate some type of probability matching coefficient, so the tenant with the highest coefficient wins.
But that introduces some type of guess mechanics (yes, this guess is educated, but nevertheless).
So the only way to clearly identify, to which tenant assign a namespace in such a situation - is to use labels. Thus we don't have to guess

**Comment by bsctl** [comment]
@MaxFedotov thanks for the explanation

> So in this situation, teamalpha-production-mycoolproject tenant obviously wins.

This is my expectation.

> But what if it is not, what the user wants?
What if he wants it to be assigned to teamalpha tenant?

He has to specify by using the label in that case. Complete match wins.

**Comment by bsctl** [comment]
> But what if it is not, what the user wants?
What if he wants it to be assigned to `teamalpha` tenant?

It's a bit unusual request .. imho.
A more common expectation is to assign the namespace `teamalpha-production-mycoolproject-web` to the tenant `teamalpha-production-mycoolproject`.

As for example, in another domain, if you have a reverse proxy with following routes:

A. http://teamalpha
B. http://teamalpha/production
C. http://teamalpha/production/mycoolproject

what's your expected matching for a request having http://teamalpha/production/mycoolproject/web ? Mine is (C). Does it make sense? However, users always can require to match a different tenant by using the label `capsule.clastix.io/tenant=teamalpha` and force the common logic.

Honestly, I prefer to have a better UX following a common accepted logic than a worse UX trying to satisfy more uncommon corner cases.

**Comment by MaxFedotov** [comment]
@prometherion @bsctl what do you think about using https://github.com/schollz/closestmatch library in order to find most suitable tenant for a namespace?

**Comment by prometherion** [comment]
I have to say that in the beginning, I was ok with the case of a user, having multiple Tenants, had to put a label to Namespace in order to assign to the right Tenant when Capsule was running with the force Tenantt prefix flag.

But I missed the point that the underlying reason for this feature is to don't let users do something like this:

```
kubectl create namespace oil-dev -o yaml > /tmp/ns.yaml
# vi /tmp/ns.yaml adding label `capsule.clastix.io/tenant=oil`
kubectl apply -f /tmp/ns.yaml
```

However, the remarks reported by @MaxFedotov are legitimated as the @bsctl ones.

Tl;dr; let's implement the _closest match_, although I'm not really sure of this approach: what I mean is that when the accuracy is not 100% we should block the creation since Capsule doesn't have to pick up decisions on behalf of the Tenant Owner and if their case is fuzzy due to multiple Tenant with prefixes (e.g., `awesome` and `awesome-org` or any case reported by Max), it's their fault and should be handled in an imperative way.

**Comment by bsctl** [comment]
> Tl;dr; let's implement the closest match, although I'm not really sure of this approach: what I mean is that when the accuracy is not 100% we should block the creation since Capsule doesn't have to pick up decisions on behalf of the Tenant Owner and if their case is fuzzy due to multiple Tenant with prefixes (e.g., awesome and awesome-org or any case reported by Max), it's their fault and should be handled in an imperative way.

+1

> But I missed the point that the underlying reason for this feature

we can ask the user to specify a label and this makes sense when the user has multiple tenants to choice and when the `--force-tenant-prefix` is set to `FALSE`.

However, when `--force-tenant-prefix` is set to `TRUE`, we already provided the user a criteria to indicate the tenant. Requiring to specify it again with a label, feels like a poor UX. It recall me a sort of government bureaucracy :)

**Comment by MaxFedotov** [comment]
@prometherion There won't be 100% match for tenant, as the namespace name contains additional words, e.g.
```
	wordsToTest := []string{"teamalpha", "teamalpha-production", "teamalpha-production-mycoolproject"}
	bagSizes := []int{2}
	cm := closestmatch.New(wordsToTest, bagSizes)
	fmt.Println(cm.Closest("teamalpha-production-mycoolproject-web"))
	fmt.Println(cm.AccuracyMutatingWords())
```
and result is 
```
teamalpha-production-mycoolproject
72.22222222222223
```

So event in this situation it's 72% match

**Comment by bsctl** [comment]
@MaxFedotov it might seem odd .. but why not just find the first match of the namespace in the sorted list of tenants? If no match, just return an error message saying "_Not able to find a tenant to match and ... bla bla bla_"

**Comment by bsctl** [comment]
@MaxFedotov not sure if we really need for fuzzy-logic here. May be a more traditional matching logic is enough. Have you looked at some router/proxy out there?

> A. http://teamalpha
B. http://teamalpha/production
C. http://teamalpha/production/mycoolproject
> what's your expected matching for a request having http://teamalpha/production/mycoolproject/web ? 

I feel they already solved our same problem with slashes `/` instead of dashes `-`

**Comment by prometherion** [comment]
> I feel they already solved our same problem with slashes / instead of dashes -

I'm missing the point here, mates: why are we talking about URL and slashes? Namespace names adhere to DNS RFC-1123 and that means no special characters, except for the dash (`-`).

Regarding the case reported by @MaxFedotov, when creating a Namespace named `teamalpha-production-mycoolproject-web` with the following Tenant list

> 1.teamalpha
> 2. teamalpha-production
> 3. teamalpha-production-mycoolproject

I'm expecting to get it assigned to `teamalpha-production-mycoolproject`.

This is an odd behavior put in place by the end-user since I honestly don't see any reason behind the naming of this kind but can be easily solved by `capsule.clastix.io/teant` label: for the said case, we're not going to print an error, but rather selecting the most suitable Tenant and if it's wrong, just need to use the label (`/shrug`).

We can easily implement this sorting the list of available tenants by string length and iterating over this, returning the first match: it's safe enough, since the label selection is performed with the Owner field indexer.

@bsctl @MaxFedotov deal?
